### PR TITLE
fix: [CDS-52386]: workflow rbac cover index

### PR DESCRIPTION
**Pull request by abhinav-harness**
### Issue in hand
In rbac calls in CG we pre-calculate all the permissions of user by getting different entities and meta information of those entities. Due to this reason we see a lot of mongo calls to get entities like workflows, env, service etc for whole account.

### What issue has it caused
Some of the customers have higher number of workflows - this results in cursor taking higher time to get all those records which shows up as slow mongo queries.

### Proposed solution
Mongo offers a way known as cover index - https://www.mongodb.com/docs/manual/indexes/#covered-queries - which basically means we need not go to disk to get records and those records are directly served from index if we have correct projections. We plan to use that for workflows. 

### Why so many changes at core layer in index manager?
Our index manager has written a common sense logic - which is not to have any composite index which has ID as part of it. As any index having ID is of no use since it in itself is a sufficient index without other nodes in index. 
To get around and make it more "informed" i have made change in fmwk to allow creating index with id in nodes (for cover index use case) by allowing developers with index with certain name to allow have id in composite index.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/43403)

### CODE REVIEW

**Review by Fernando-Dourado — DISMISSED** [review]
> Reviewed 8 of 8 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43403)* status: 8 file(s) to review. 1 of 2 approvals obtained.

**Review by ankit-harness — DISMISSED** [review]
:lgtm:

> Reviewed 5 of 8 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43403#-:-NNSC64tA0GxpZcyCYlG:bnfp4nl)* status: 3 file(s) to review. 1 of 2 approvals obtained.

File: 400-rest/src/main/java/software/wings/service/impl/workflow/WorkflowServiceImpl.java [hunk 1]
```diff
@@ -459,11 +460,15 @@ public PageResponse<StateMachine> listStateMachines(PageRequest<StateMachine> re
...
   @Override
-  public List<Workflow> list(String accountId, List<String> projectFields) {
+  public List<Workflow> list(String accountId, List<String> projectFields, String queryHint) {
     Query<Workflow> workflowQuery =
         wingsPersistence.createQuery(Workflow.class).filter(WorkflowKeys.accountId, accountId);
     emptyIfNull(projectFields).forEach(field -> { workflowQuery.project(field, true); });
-    return emptyIfNull(workflowQuery.asList());
+    FindOptions findOptions = new FindOptions();
+    if (isNotEmpty(queryHint)) {
+      findOptions.hint(BasicDBUtils.getIndexObject(WorkflowExecution.mongoIndexes(), queryHint));
```
Comment (kshitij-harness) [inline 400-rest/src/main/java/software/wings/service/impl/workflow/WorkflowServiceImpl.java:469]: This should be Workflow.mongoIndexes()

**Review by kshitij-harness — CHANGES_REQUESTED** [review]

**Review by kshitij-harness — APPROVED** [review]
> Reviewed 3 of 8 files at r3, 3 of 3 files at r4, 2 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43403)* status: 3 file(s) to review. 1 of 2 approvals obtained.

**Comment by rijajoo** [comment]
"..by allowing developers with index with certain name to allow have id in composite index."
Don't we allow indexes with id in composite index to have any names here?

**Review by RajBaviskar — APPROVED** [review]

**Comment by abhinav-harness** [comment]
@rijajoo thats bad english :P 
the aim was to not have uuid as part of any composite index

**Review by RajBaviskar — APPROVED** [review]
> Reviewed 3 of 8 files at r3, 3 of 3 files at r4, 2 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43403)* status: :shipit: complete! ✓

**Review by RajBaviskar — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43403)* status: :shipit: complete! ✓

**Pull request merged**